[BUG] Cannot load AASX files with JSON payload

Hi all

i fail to load aasx files with json payload into the server.

`
2024-10-02T14:05:56.462Z  INFO 1 --- [           main] a.p.AasEnvironmentPreconfigurationLoader : Loading AAS Environment (1/3) from file 'jsonAO2000.aasx'
2024-10-02T14:05:56.697Z  WARN 1 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'preconfigurationLoaderInitializer' defined in URL [jar:file:/application/basyxExecutable.jar!/BOOT-INF/lib/basyx.aasenvironment-core-2.0.0-SNAPSHOT.jar!/org/eclipse/digitaltwin/basyx/aasenvironment/preconfiguration/PreconfigurationLoaderInitializer.class]: deserialization failed
2024-10-02T14:05:56.705Z  WARN 1 --- [           main] o.s.b.f.support.DisposableBeanAdapter    : Invocation of close method failed on bean with name 'mqttClient': Client is connected (32100)
2024-10-02T14:05:56.725Z  INFO 1 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
2024-10-02T14:05:56.770Z  INFO 1 --- [           main] .s.b.a.l.ConditionEvaluationReportLogger : 
Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
2024-10-02T14:05:56.819Z ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'preconfigurationLoaderInitializer' defined in URL [jar:file:/application/basyxExecutable.jar!/BOOT-INF/lib/basyx.aasenvironment-core-2.0.0-SNAPSHOT.jar!/org/eclipse/digitaltwin/basyx/aasenvironment/preconfiguration/PreconfigurationLoaderInitializer.class]: deserialization failed
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1770) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:598) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:520) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:325) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:323) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:199) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:973) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:942) ~[spring-context-6.0.12.jar!/:6.0.12]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:608) ~[spring-context-6.0.12.jar!/:6.0.12]
        at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.1.4.jar!/:3.1.4]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:737) ~[spring-boot-3.1.4.jar!/:3.1.4]
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:439) ~[spring-boot-3.1.4.jar!/:3.1.4]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) ~[spring-boot-3.1.4.jar!/:3.1.4]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1309) ~[spring-boot-3.1.4.jar!/:3.1.4]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1298) ~[spring-boot-3.1.4.jar!/:3.1.4]
        at org.eclipse.digitaltwin.basyx.aasenvironment.component.AasEnvironmentComponent.main(AasEnvironmentComponent.java:42) ~[classes!/:2.0.0-SNAPSHOT]
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
        at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) ~[basyxExecutable.jar:2.0.0-SNAPSHOT]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:95) ~[basyxExecutable.jar:2.0.0-SNAPSHOT]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) ~[basyxExecutable.jar:2.0.0-SNAPSHOT]
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65) ~[basyxExecutable.jar:2.0.0-SNAPSHOT]
Caused by: org.eclipse.digitaltwin.aas4j.v3.dataformat.core.DeserializationException: deserialization failed
        at org.eclipse.digitaltwin.aas4j.v3.dataformat.xml.XmlDeserializer.read(XmlDeserializer.java:120) ~[aas4j-dataformat-xml-1.0.2.jar!/:na]
        at org.eclipse.digitaltwin.aas4j.v3.dataformat.aasx.AASXDeserializer.read(AASXDeserializer.java:99) ~[aas4j-dataformat-aasx-1.0.2.jar!/:na]
        at org.eclipse.digitaltwin.aas4j.v3.dataformat.aasx.AASXDeserializer.parseReferencedFilePathsFromAASX(AASXDeserializer.java:201) ~[aas4j-dataformat-aasx-1.0.2.jar!/:na]
        at org.eclipse.digitaltwin.aas4j.v3.dataformat.aasx.AASXDeserializer.getRelatedFiles(AASXDeserializer.java:125) ~[aas4j-dataformat-aasx-1.0.2.jar!/:na]
        at org.eclipse.digitaltwin.basyx.aasenvironment.environmentloader.CompleteEnvironment.fromInputStream(CompleteEnvironment.java:113) ~[basyx.aasenvironment-core-2.0.0-SNAPSHOT.jar!/:2.0.0-SNAPSHOT]
        at org.eclipse.digitaltwin.basyx.aasenvironment.environmentloader.CompleteEnvironment.fromFile(CompleteEnvironment.java:96) ~[basyx.aasenvironment-core-2.0.0-SNAPSHOT.jar!/:2.0.0-SNAPSHOT]
        at org.eclipse.digitaltwin.basyx.aasenvironment.preconfiguration.AasEnvironmentPreconfigurationLoader.loadPreconfiguredEnvironments(AasEnvironmentPreconfigurationLoader.java:85) ~[basyx.aasenvironment-core-2.0.0-SNAPSHOT.jar!/:2.0.0-SNAPSHOT]
        at org.eclipse.digitaltwin.basyx.aasenvironment.preconfiguration.PreconfigurationLoaderInitializer.loadPreconfiguredEnvironment(PreconfigurationLoaderInitializer.java:68) ~[basyx.aasenvironment-core-2.0.0-SNAPSHOT.jar!/:2.0.0-SNAPSHOT]
        at org.eclipse.digitaltwin.basyx.aasenvironment.preconfiguration.PreconfigurationLoaderInitializer.afterPropertiesSet(PreconfigurationLoaderInitializer.java:60) ~[basyx.aasenvironment-core-2.0.0-SNAPSHOT.jar!/:2.0.0-SNAPSHOT]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1817) ~[spring-beans-6.0.12.jar!/:6.0.12]
        at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1766) ~[spring-beans-6.0.12.jar!/:6.0.12]
        ... 24 common frames omitted
Caused by: com.fasterxml.jackson.core.JsonParseException: Unexpected character '{' (code 123) in prolog; expected '<'
at [row,col {unknown-source}]: [1,1]
        at com.fasterxml.jackson.dataformat.xml.XmlFactory._initializeXmlReader(XmlFactory.java:734) ~[jackson-dataformat-xml-2.15.2.jar!/:2.15.2]
        at com.fasterxml.jackson.dataformat.xml.XmlFactory._createParser(XmlFactory.java:618) ~[jackson-dataformat-xml-2.15.2.jar!/:2.15.2]
        at com.fasterxml.jackson.dataformat.xml.XmlFactory.createParser(XmlFactory.java:496) ~[jackson-dataformat-xml-2.15.2.jar!/:2.15.2]
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3772) ~[jackson-databind-2.15.2.jar!/:2.15.2]
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3740) ~[jackson-databind-2.15.2.jar!/:2.15.2]
        at org.eclipse.digitaltwin.aas4j.v3.dataformat.xml.XmlDeserializer.read(XmlDeserializer.java:118) ~[aas4j-dataformat-xml-1.0.2.jar!/:na]
        ... 34 common frames omitted
Caused by: com.ctc.wstx.exc.WstxUnexpectedCharException: Unexpected character '{' (code 123) in prolog; expected '<'
at [row,col {unknown-source}]: [1,1]
        at com.ctc.wstx.sr.StreamScanner.throwUnexpectedChar(StreamScanner.java:666) ~[woodstox-core-6.5.1.jar!/:6.5.1]
        at com.ctc.wstx.sr.BasicStreamReader.nextFromProlog(BasicStreamReader.java:2148) ~[woodstox-core-6.5.1.jar!/:6.5.1]
        at com.ctc.wstx.sr.BasicStreamReader.next(BasicStreamReader.java:1180) ~[woodstox-core-6.5.1.jar!/:6.5.1]
        at com.fasterxml.jackson.dataformat.xml.XmlFactory._initializeXmlReader(XmlFactory.java:729) ~[jackson-dataformat-xml-2.15.2.jar!/:2.15.2]
`
Aas4j should support this https://github.com/eclipse-aas4j/aas4j/issues/287 feature.

I also attach a demo AASX file (please rename to aasx) [jsonAO2000docs.zip](https://github.com/user-attachments/files/17231565/jsonAO2000docs.zip)

Thanks!

--- a/basyx.aasenvironment/basyx.aasenvironment-client/src/main/java/org/eclipse/digitaltwin/basyx/aasenvironment/client/ConnectedAasManager.java
+++ b/basyx.aasenvironment/basyx.aasenvironment-client/src/main/java/org/eclipse/digitaltwin/basyx/aasenvironment/client/ConnectedAasManager.java
@@ -223,10 +223,7 @@ public class ConnectedAasManager {
 			throw new RegistryHttpRequestException(aasIdentifier, e);
 		}
 
-		Reference smRef = AasUtils.toReference(AasUtils.toReference(shell), submodel);
-
-		// TODO See https://github.com/eclipse-aas4j/aas4j/issues/308
-		smRef.setReferredSemanticId(submodel.getSemanticId());
+		Reference smRef = AasUtils.toReference(AasUtils.toReference(shell), submodel, true);
 
 		aasRepository.addSubmodelReference(aasIdentifier, smRef);
 	}

--- a/basyx.submodelrepository/basyx.submodelrepository-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/SubmodelRepositoryApiHTTPController.java
+++ b/basyx.submodelrepository/basyx.submodelrepository-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/SubmodelRepositoryApiHTTPController.java
@@ -285,6 +285,10 @@ public class SubmodelRepositoryApiHTTPController implements SubmodelRepositoryHT
 
 	@Override
 	public ResponseEntity<OperationResult> invokeOperationSubmodelRepo(Base64UrlEncodedIdentifier submodelIdentifier, String idShortPath, @Valid OperationRequest body, @Valid Boolean async) {
+
+		// TODO: #566 Add async operation execution support to
+		// SubmodelRepositoryController
+
 		Invokable invokable = inArgs -> repository.invokeOperation(submodelIdentifier.getIdentifier(), idShortPath, inArgs);
 		return ResponseEntity.ok(OperationRequestExecutor.executeOperationRequestSynchronously(invokable, body));
 	}

--- a/basyx.submodelrepository/basyx.submodelrepository-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/SubmodelRepositoryApiHTTPController.java
+++ b/basyx.submodelrepository/basyx.submodelrepository-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/http/SubmodelRepositoryApiHTTPController.java
@@ -48,6 +48,8 @@ import org.eclipse.digitaltwin.basyx.http.Base64UrlEncoder;
 import org.eclipse.digitaltwin.basyx.http.pagination.Base64UrlEncodedCursor;
 import org.eclipse.digitaltwin.basyx.http.pagination.PagedResult;
 import org.eclipse.digitaltwin.basyx.http.pagination.PagedResultPagingMetadata;
+import org.eclipse.digitaltwin.basyx.operation.OperationRequestExecutor;
+import org.eclipse.digitaltwin.basyx.operation.OperationRequestExecutor.Invokable;
 import org.eclipse.digitaltwin.basyx.pagination.GetSubmodelElementsResult;
 import org.eclipse.digitaltwin.basyx.submodelrepository.SubmodelRepository;
 import org.eclipse.digitaltwin.basyx.submodelrepository.http.pagination.GetSubmodelsResult;
@@ -283,28 +285,8 @@ public class SubmodelRepositoryApiHTTPController implements SubmodelRepositoryHT
 
 	@Override
 	public ResponseEntity<OperationResult> invokeOperationSubmodelRepo(Base64UrlEncodedIdentifier submodelIdentifier, String idShortPath, @Valid OperationRequest body, @Valid Boolean async) {
-		List<OperationVariable> inVars = new ArrayList<>();
-		inVars.addAll(body.getInputArguments());
-		inVars.addAll(body.getInoutputArguments());
-
-		List<OperationVariable> result = Arrays.asList(repository.invokeOperation(submodelIdentifier.getIdentifier(), idShortPath, inVars.toArray(new OperationVariable[0])));
-
-		List<OperationVariable> outVars = new ArrayList<>(result);
-		List<OperationVariable> inoutputVars = new ArrayList<>();
-
-		if (!body.getInoutputArguments().isEmpty()) {
-			List<String> inoutputVarsIdShorts = body.getInoutputArguments().stream().map(OperationVariable::getValue).map(SubmodelElement::getIdShort).toList();
-
-			inoutputVars = result.stream().filter(opVar -> inoutputVarsIdShorts.contains(opVar.getValue().getIdShort())).toList();
-
-			outVars.removeAll(inoutputVars);
-		}
-
-		return ResponseEntity.ok(createOperationResult(outVars, inoutputVars));
-	}
-
-	private OperationResult createOperationResult(List<OperationVariable> outputVars, List<OperationVariable> inoutputVars) {
-		return new DefaultOperationResult.Builder().outputArguments(outputVars).inoutputArguments(inoutputVars).build();
+		Invokable invokable = inArgs -> repository.invokeOperation(submodelIdentifier.getIdentifier(), idShortPath, inArgs);
+		return ResponseEntity.ok(OperationRequestExecutor.executeOperationRequestSynchronously(invokable, body));
 	}
 
 	private String getEncodedCursorFromCursorResult(CursorResult<?> cursorResult) {

--- a/basyx.submodelservice/basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/operation/OperationRequestExecutor.java
+++ b/basyx.submodelservice/basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/operation/OperationRequestExecutor.java
@@ -1,0 +1,79 @@
+/*******************************************************************************
+ * Copyright (C) 2023 the Eclipse BaSyx Authors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * SPDX-License-Identifier: MIT
+ ******************************************************************************/
+
+package org.eclipse.digitaltwin.basyx.operation;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.digitaltwin.aas4j.v3.model.OperationRequest;
+import org.eclipse.digitaltwin.aas4j.v3.model.OperationResult;
+import org.eclipse.digitaltwin.aas4j.v3.model.OperationVariable;
+import org.eclipse.digitaltwin.aas4j.v3.model.SubmodelElement;
+import org.eclipse.digitaltwin.aas4j.v3.model.impl.DefaultOperationResult;
+import org.eclipse.digitaltwin.basyx.submodelservice.SubmodelService;
+
+/**
+ * Executes an Operation Request
+ * 
+ * @author mateusmolina
+ */
+public class OperationRequestExecutor {
+
+    private OperationRequestExecutor() {
+    }
+
+    public static OperationResult executeOperationRequestSynchronously(Invokable invokable, OperationRequest request) {
+        List<OperationVariable> inVars = new ArrayList<>();
+        inVars.addAll(request.getInputArguments());
+        inVars.addAll(request.getInoutputArguments());
+
+        List<OperationVariable> result = Arrays.asList(invokable.invoke(inVars.toArray(new OperationVariable[0])));
+
+        List<OperationVariable> outVars = new ArrayList<>(result);
+        List<OperationVariable> inoutputVars = new ArrayList<>();
+
+        if (!request.getInoutputArguments().isEmpty()) {
+            List<String> inoutputVarsIdShorts = request.getInoutputArguments().stream().map(OperationVariable::getValue).map(SubmodelElement::getIdShort).toList();
+
+            inoutputVars = result.stream().filter(opVar -> inoutputVarsIdShorts.contains(opVar.getValue().getIdShort())).toList();
+
+            outVars.removeAll(inoutputVars);
+        }
+
+        return createSuccessOperationResult(outVars, inoutputVars);
+    }
+
+    private static OperationResult createSuccessOperationResult(List<OperationVariable> outputVars, List<OperationVariable> inoutputVars) {
+        return new DefaultOperationResult.Builder().success(true).outputArguments(outputVars).inoutputArguments(inoutputVars).build();
+    }
+
+    @FunctionalInterface
+    public interface Invokable {
+        OperationVariable[] invoke(OperationVariable[] arguments);
+    }
+
+}

--- a/basyx.submodelservice/basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/operation/OperationRequestExecutor.java
+++ b/basyx.submodelservice/basyx.submodelservice-core/src/main/java/org/eclipse/digitaltwin/basyx/operation/OperationRequestExecutor.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (C) 2023 the Eclipse BaSyx Authors
+ * Copyright (C) 2024 the Eclipse BaSyx Authors
  *
  * Permission is hereby granted, free of charge, to any person obtaining
  * a copy of this software and associated documentation files (the
@@ -34,7 +34,6 @@ import org.eclipse.digitaltwin.aas4j.v3.model.OperationResult;
 import org.eclipse.digitaltwin.aas4j.v3.model.OperationVariable;
 import org.eclipse.digitaltwin.aas4j.v3.model.SubmodelElement;
 import org.eclipse.digitaltwin.aas4j.v3.model.impl.DefaultOperationResult;
-import org.eclipse.digitaltwin.basyx.submodelservice.SubmodelService;
 
 /**
  * Executes an Operation Request

--- a/basyx.submodelservice/basyx.submodelservice-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/http/SubmodelServiceHTTPApiController.java
+++ b/basyx.submodelservice/basyx.submodelservice-http/src/main/java/org/eclipse/digitaltwin/basyx/submodelservice/http/SubmodelServiceHTTPApiController.java
@@ -27,16 +27,12 @@ package org.eclipse.digitaltwin.basyx.submodelservice.http;
 
 import java.io.IOException;
 import java.io.InputStream;
-import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.List;
 
 import org.eclipse.digitaltwin.aas4j.v3.model.OperationRequest;
 import org.eclipse.digitaltwin.aas4j.v3.model.OperationResult;
-import org.eclipse.digitaltwin.aas4j.v3.model.OperationVariable;
 import org.eclipse.digitaltwin.aas4j.v3.model.Submodel;
 import org.eclipse.digitaltwin.aas4j.v3.model.SubmodelElement;
-import org.eclipse.digitaltwin.aas4j.v3.model.impl.DefaultOperationResult;
 import org.eclipse.digitaltwin.basyx.core.exceptions.ElementDoesNotExistException;
 import org.eclipse.digitaltwin.basyx.core.exceptions.ElementNotAFileException;
 import org.eclipse.digitaltwin.basyx.core.exceptions.FileDoesNotExistException;
@@ -45,6 +41,8 @@ import org.eclipse.digitaltwin.basyx.core.pagination.PaginationInfo;
 import org.eclipse.digitaltwin.basyx.http.pagination.Base64UrlEncodedCursor;
 import org.eclipse.digitaltwin.basyx.http.pagination.PagedResult;
 import org.eclipse.digitaltwin.basyx.http.pagination.PagedResultPagingMetadata;
+import org.eclipse.digitaltwin.basyx.operation.OperationRequestExecutor;
+import org.eclipse.digitaltwin.basyx.operation.OperationRequestExecutor.Invokable;
 import org.eclipse.digitaltwin.basyx.pagination.GetSubmodelElementsResult;
 import org.eclipse.digitaltwin.basyx.serialization.SubmodelMetadataUtil;
 import org.eclipse.digitaltwin.basyx.submodelservice.SubmodelService;
@@ -236,30 +234,9 @@ public class SubmodelServiceHTTPApiController implements SubmodelServiceHTTPApi 
 	public ResponseEntity<OperationResult> invokeOperation(
 			@Parameter(in = ParameterIn.PATH, description = "IdShort path to the submodel element (dot-separated)", required = true, schema = @Schema()) @PathVariable("idShortPath") String idShortPath,
 			@Parameter(in = ParameterIn.DEFAULT, description = "Operation request object", required = true, schema = @Schema()) @Valid @RequestBody OperationRequest body) {
-		List<OperationVariable> inVars = new ArrayList<>();
-		inVars.addAll(body.getInputArguments());
-		inVars.addAll(body.getInoutputArguments());
 
-		List<OperationVariable> result = Arrays.asList(service.invokeOperation(idShortPath, inVars.toArray(new OperationVariable[0])));
-
-		List<OperationVariable> outVars = new ArrayList<>(result);
-		List<OperationVariable> inoutputVars = new ArrayList<>();
-
-		if (!body.getInoutputArguments().isEmpty()) {
-			List<String> inoutputVarsIdShorts = body.getInoutputArguments().stream().map(OperationVariable::getValue).map(SubmodelElement::getIdShort).toList();
-
-			inoutputVars = result.stream().filter(opVar -> inoutputVarsIdShorts.contains(opVar.getValue().getIdShort())).toList();
-
-			outVars.removeAll(inoutputVars);
-		}
-
-		return ResponseEntity.ok(createOperationResult(outVars, inoutputVars));
-	}
-
-	private OperationResult createOperationResult(List<OperationVariable> outputVars, List<OperationVariable> inoutputVars) {
-		return new DefaultOperationResult.Builder()
-				.outputArguments(outputVars).inoutputArguments(inoutputVars)
-				.build();
+		Invokable invokable = inArgs -> service.invokeOperation(idShortPath, inArgs);
+		return ResponseEntity.ok(OperationRequestExecutor.executeOperationRequestSynchronously(invokable, body));
 	}
 
 	private String getEncodedCursorFromCursorResult(CursorResult<?> cursorResult) {

--- a/basyx.submodelservice/basyx.submodelservice-http/src/test/resources/operation/result-inout.json
+++ b/basyx.submodelservice/basyx.submodelservice-http/src/test/resources/operation/result-inout.json
@@ -1,4 +1,5 @@
 {
+	"success": true,
 	"inoutputArguments": [
 		{
 			"value": {

--- a/basyx.submodelservice/basyx.submodelservice-http/src/test/resources/operation/result.json
+++ b/basyx.submodelservice/basyx.submodelservice-http/src/test/resources/operation/result.json
@@ -1,4 +1,5 @@
 {
+	"success": true,
 	"outputArguments": [
 		{
 		    "value": {

--- a/pom.xml
+++ b/pom.xml
@@ -79,7 +79,7 @@
 		<docker.target.platforms>linux/amd64, linux/arm64, linux/arm/v7</docker.target.platforms>
 		<docker.host.port>8081</docker.host.port>
 		<docker.container.port>8081</docker.container.port>
-		<aas4j-version>1.0.2</aas4j-version>
+		<aas4j-version>1.0.3</aas4j-version>
 	</properties>
 
 	<pluginRepositories>

